File: cpp_merge/common/test-tools.test.cpp
#include "test-tools.test.h"

void Main(void)
{
    describe(std:("Command Parsing"), [=]() mutable
    {
        describe(std:("Signature formats"), [=]() mutable
        {
            it(std:("should handle default format with no signature"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("test_cmd")}, 
                    object::pair{std:("docstring"), std:("A test command")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("arg1")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("First argument")}, 
                        object::pair{std:("required"), true}
                    }), std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("arg2")}, 
                        object::pair{std:("type"), std:("integer")}, 
                        object::pair{std:("description"), std:("Second argument")}, 
                        object::pair{std:("required"), false}
                    }) }}
                });
                expect(command->get_invokeFormat())->toBe(std:("test_cmd {arg1} {arg2} "));
            }
            );
            it(std:("should handle angle brackets"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("goto")}, 
                    object::pair{std:("signature"), std:("goto <line_number>")}, 
                    object::pair{std:("docstring"), std:("moves the window to show line_number")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("line_number")}, 
                        object::pair{std:("type"), std:("integer")}, 
                        object::pair{std:("description"), std:("line number")}, 
                        object::pair{std:("required"), true}
                    }) }}
                });
                expect(command->get_invokeFormat())->toBe(std:("goto {line_number}"));
            }
            );
            it(std:("should handle optional brackets"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("open")}, 
                    object::pair{std:("signature"), std:("open "<path>" [<line_number>]")}, 
                    object::pair{std:("docstring"), std:("opens file at path")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("path")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("file path")}, 
                        object::pair{std:("required"), true}
                    }), std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("line_number")}, 
                        object::pair{std:("type"), std:("integer")}, 
                        object::pair{std:("description"), std:("line number")}, 
                        object::pair{std:("required"), false}
                    }) }}
                });
                expect(command->get_invokeFormat())->toBe(std:("open "{path}" {line_number}"));
            }
            );
            it(std:("should handle flag-style arguments"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("grep")}, 
                    object::pair{std:("signature"), std:("grep --pattern <pattern> --file <file>")}, 
                    object::pair{std:("docstring"), std:("search for pattern in file")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("pattern")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("search pattern")}, 
                        object::pair{std:("required"), true}
                    }), std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("file")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("file to search")}, 
                        object::pair{std:("required"), true}
                    }) }}
                });
                expect(command->get_invokeFormat())->toBe(std:("grep --pattern {pattern} --file {file}"));
            }
            );
            it(std:("should handle no arguments"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("scroll_up")}, 
                    object::pair{std:("signature"), std:("scroll_up")}, 
                    object::pair{std:("docstring"), std:("scrolls up")}, 
                    object::pair{std:("arguments"), array<any>()}
                });
                expect(command->get_invokeFormat())->toBe(std:("scroll_up"));
            }
            );
        }
        );
        describe(std:("Argument validation"), [=]() mutable
        {
            it(std:("should require required arguments before optional ones"), [=]() mutable
            {
                expect([=]() mutable
                {
                    std::make_shared<Command>(object{
                        object::pair{std:("name"), std:("bad_order")}, 
                        object::pair{std:("docstring"), std:("bad argument order")}, 
                        object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("optional")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("optional")}, 
                            object::pair{std:("required"), false}
                        }), std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("required")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("required")}, 
                            object::pair{std:("required"), true}
                        }) }}
                    });
                }
                )->toThrow((new RegExp(std:("Required argument.*cannot come after optional argument"))));
            }
            );
            it(std:("should not allow duplicate argument names"), [=]() mutable
            {
                expect([=]() mutable
                {
                    std::make_shared<Command>(object{
                        object::pair{std:("name"), std:("duplicate")}, 
                        object::pair{std:("docstring"), std:("duplicate args")}, 
                        object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("arg1")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("first")}, 
                            object::pair{std:("required"), true}
                        }), std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("arg1")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("duplicate")}, 
                            object::pair{std:("required"), true}
                        }) }}
                    });
                }
                )->toThrow((new RegExp(std:("Duplicate argument nam"))));
            }
            );
        }
        );
        describe(std:("Argument name patterns"), [=]() mutable
        {
            it(std:("should accept valid argument names"), [=]() mutable
            {
                auto validNames = array<string>{ std:("a"), std:("x"), std:("n"), std:("simple"), std:("with_underscore"), std:("with-dash"), std:("with123numbers"), std:("_starts_with_underscore") };
                validNames->forEach([=](auto name) mutable
                {
                    auto command = std::make_shared<Command>(object{
                        object::pair{std:("name"), std:("test")}, 
                        object::pair{std:("docstring"), std:("test")}, 
                        object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("name")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("test")}, 
                            object::pair{std:("required"), true}
                        }) }}
                    });
                    expect(const_(command->arguments)[0]->name)->toBe(name);
                }
                );
            }
            );
            it(std:("should reject invalid argument names"), [=]() mutable
            {
                auto invalidNames = array<string>{ std:("123starts_with_number"), string_empty };
                invalidNames->forEach([=](auto name) mutable
                {
                    expect([=]() mutable
                    {
                        std::make_shared<Command>(object{
                            object::pair{std:("name"), std:("test")}, 
                            object::pair{std:("docstring"), std:("test")}, 
                            object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                                object::pair{std:("name"), std:("name")}, 
                                object::pair{std:("type"), std:("string")}, 
                                object::pair{std:("description"), std:("test")}, 
                                object::pair{std:("required"), true}
                            }) }}
                        });
                    }
                    )->toThrow((new RegExp(std:("Invalid argument nam"))));
                }
                );
            }
            );
        }
        );
        describe(std:("Signature and argument consistency"), [=]() mutable
        {
            it(std:("should require all arguments to appear in signature"), [=]() mutable
            {
                expect([=]() mutable
                {
                    std::make_shared<Command>(object{
                        object::pair{std:("name"), std:("missing_arg")}, 
                        object::pair{std:("signature"), std:("missing_arg <existing_arg>")}, 
                        object::pair{std:("docstring"), std:("missing argument in signature")}, 
                        object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("existing_arg")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("exists")}, 
                            object::pair{std:("required"), true}
                        }), std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("missing_arg")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("not in signature")}, 
                            object::pair{std:("required"), true}
                        }) }}
                    });
                }
                )->toThrow((new RegExp(std:("Missing argument.*in signatur"))));
            }
            );
            it(std:("should reject extra arguments in signature"), [=]() mutable
            {
                expect([=]() mutable
                {
                    std::make_shared<Command>(object{
                        object::pair{std:("name"), std:("extra_arg")}, 
                        object::pair{std:("signature"), std:("extra_arg <arg1> <extra>")}, 
                        object::pair{std:("docstring"), std:("extra argument in signature")}, 
                        object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                            object::pair{std:("name"), std:("arg1")}, 
                            object::pair{std:("type"), std:("string")}, 
                            object::pair{std:("description"), std:("exists")}, 
                            object::pair{std:("required"), true}
                        }) }}
                    });
                }
                )->toThrow((new RegExp(std:("Argument names.*do not matc"))));
            }
            );
        }
        );
        describe(std:("Function calling tool generation"), [=]() mutable
        {
            it(std:("should generate OpenAI function calling tool"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("test_function")}, 
                    object::pair{std:("docstring"), std:("A test function for OpenAI")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("required_arg")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("Required string argument")}, 
                        object::pair{std:("required"), true}
                    }), std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("enum_arg")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("Enum argument")}, 
                        object::pair{std:("required"), true}, 
                        object::pair{std:("enum"), array<string>{ std:("option1"), std:("option2") }}
                    }), std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("optional_arg")}, 
                        object::pair{std:("type"), std:("integer")}, 
                        object::pair{std:("description"), std:("Optional integer argument")}, 
                        object::pair{std:("required"), false}
                    }) }}
                });
                auto tool = command->getFunctionCallingTool();
                expect(tool->type)->toBe(std:("function"));
                expect(tool->function["name"])->toBe(std:("test_function"));
                expect(tool->function["description"])->toBe(std:("A test function for OpenAI"));
                auto properties = tool->function["parameters"]["properties"];
                expect(properties["required_arg"]["type"])->toBe(std:("string"));
                expect(properties["optional_arg"]["type"])->toBe(std:("integer"));
                expect(properties["enum_arg"]["enum"])->toEqual(array<string>{ std:("option1"), std:("option2") });
                auto required = tool->function["parameters"]["required"];
                expect(required)->toContain(std:("required_arg"));
                expect(required)->toContain(std:("enum_arg"));
                expect(required)->not->toContain(std:("optional_arg"));
            }
            );
        }
        );
        describe(std:("Multi-line commands"), [=]() mutable
        {
            it(std:("should handle commands with end markers"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("edit")}, 
                    object::pair{std:("signature"), std:("edit <filename>")}, 
                    object::pair{std:("docstring"), std:("Edit a file with multi-line content")}, 
                    object::pair{std:("endName"), std:("EOF")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("filename")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("file to edit")}, 
                        object::pair{std:("required"), true}
                    }) }}
                });
                expect(command->get_invokeFormat())->toBe(std:("edit {filename}"));
                expect(command->endName)->toBe(std:("EOF"));
            }
            );
        }
        );
        describe(std:("Custom argument format"), [=]() mutable
        {
            it(std:("should handle custom argument formatting"), [=]() mutable
            {
                auto command = std::make_shared<Command>(object{
                    object::pair{std:("name"), std:("custom_format")}, 
                    object::pair{std:("docstring"), std:("Test custom argument formatting")}, 
                    object::pair{std:("arguments"), array<std::shared_ptr<ArgumentImpl>>{ std::make_shared<Argument>(object{
                        object::pair{std:("name"), std:("arg1")}, 
                        object::pair{std:("type"), std:("string")}, 
                        object::pair{std:("description"), std:("Custom formatted argument")}, 
                        object::pair{std:("required"), true}, 
                        object::pair{std:("argumentFormat"), std:("--{value}")}
                    }) }}
                });
                expect(const_(command->arguments)[0]->argumentFormat)->toBe(std:("--{value}"));
                expect(command->get_invokeFormat())->toBe(std:("custom_format {arg1} "));
            }
            );
        }
        );
    }
    );
}

MAIN

File: SWEagent/tests/test-tools.test.ts
/**
 * Tools tests converted from test_tools_command_parsing.py
 */

import { describe, it, expect } from '@jest/globals';
import { Command, Argument } from '../src/tools/commands';

describe('Command Parsing', () => {
  describe('Signature formats', () => {
    it('should handle default format with no signature', () => {
      const command = new Command({
        name: 'test_cmd',
        docstring: 'A test command',
        arguments: [
          new Argument({
            name: 'arg1',
            type: 'string',
            description: 'First argument',
            required: true,
          }),
          new Argument({
            name: 'arg2',
            type: 'integer',
            description: 'Second argument',
            required: false,
          }),
        ],
      });

      expect(command.invokeFormat).toBe('test_cmd {arg1} {arg2} ');
    });

    it('should handle angle brackets', () => {
      const command = new Command({
        name: 'goto',
        signature: 'goto <line_number>',
        docstring: 'moves the window to show line_number',
        arguments: [
          new Argument({
            name: 'line_number',
            type: 'integer',
            description: 'line number',
            required: true,
          }),
        ],
      });

      expect(command.invokeFormat).toBe('goto {line_number}');
    });

    it('should handle optional brackets', () => {
      const command = new Command({
        name: 'open',
        signature: 'open "<path>" [<line_number>]',
        docstring: 'opens file at path',
        arguments: [
          new Argument({
            name: 'path',
            type: 'string',
            description: 'file path',
            required: true,
          }),
          new Argument({
            name: 'line_number',
            type: 'integer',
            description: 'line number',
            required: false,
          }),
        ],
      });

      expect(command.invokeFormat).toBe('open "{path}" {line_number}');
    });

    it('should handle flag-style arguments', () => {
      const command = new Command({
        name: 'grep',
        signature: 'grep --pattern <pattern> --file <file>',
        docstring: 'search for pattern in file',
        arguments: [
          new Argument({
            name: 'pattern',
            type: 'string',
            description: 'search pattern',
            required: true,
          }),
          new Argument({
            name: 'file',
            type: 'string',
            description: 'file to search',
            required: true,
          }),
        ],
      });

      expect(command.invokeFormat).toBe('grep --pattern {pattern} --file {file}');
    });

    it('should handle no arguments', () => {
      const command = new Command({
        name: 'scroll_up',
        signature: 'scroll_up',
        docstring: 'scrolls up',
        arguments: [],
      });

      expect(command.invokeFormat).toBe('scroll_up');
    });
  });

  describe('Argument validation', () => {
    it('should require required arguments before optional ones', () => {
      expect(() => {
        new Command({
          name: 'bad_order',
          docstring: 'bad argument order',
          arguments: [
            new Argument({
              name: 'optional',
              type: 'string',
              description: 'optional',
              required: false,
            }),
            new Argument({
              name: 'required',
              type: 'string',
              description: 'required',
              required: true,
            }),
          ],
        });
      }).toThrow(/Required argument.*cannot come after optional arguments/);
    });

    it('should not allow duplicate argument names', () => {
      expect(() => {
        new Command({
          name: 'duplicate',
          docstring: 'duplicate args',
          arguments: [
            new Argument({
              name: 'arg1',
              type: 'string',
              description: 'first',
              required: true,
            }),
            new Argument({
              name: 'arg1',
              type: 'string',
              description: 'duplicate',
              required: true,
            }),
          ],
        });
      }).toThrow(/Duplicate argument name/);
    });
  });

  describe('Argument name patterns', () => {
    it('should accept valid argument names', () => {
      const validNames = [
        'a',
        'x',
        'n',
        'simple',
        'with_underscore',
        'with-dash',
        'with123numbers',
        '_starts_with_underscore',
      ];

      validNames.forEach((name) => {
        const command = new Command({
          name: 'test',
          docstring: 'test',
          arguments: [
            new Argument({
              name,
              type: 'string',
              description: 'test',
              required: true,
            }),
          ],
        });

        expect(command.arguments[0].name).toBe(name);
      });
    });

    it('should reject invalid argument names', () => {
      const invalidNames = ['123starts_with_number', ''];

      invalidNames.forEach((name) => {
        expect(() => {
          new Command({
            name: 'test',
            docstring: 'test',
            arguments: [
              new Argument({
                name,
                type: 'string',
                description: 'test',
                required: true,
              }),
            ],
          });
        }).toThrow(/Invalid argument name/);
      });
    });
  });

  describe('Signature and argument consistency', () => {
    it('should require all arguments to appear in signature', () => {
      expect(() => {
        new Command({
          name: 'missing_arg',
          signature: 'missing_arg <existing_arg>',
          docstring: 'missing argument in signature',
          arguments: [
            new Argument({
              name: 'existing_arg',
              type: 'string',
              description: 'exists',
              required: true,
            }),
            new Argument({
              name: 'missing_arg',
              type: 'string',
              description: 'not in signature',
              required: true,
            }),
          ],
        });
      }).toThrow(/Missing argument.*in signature/);
    });

    it('should reject extra arguments in signature', () => {
      expect(() => {
        new Command({
          name: 'extra_arg',
          signature: 'extra_arg <arg1> <extra>',
          docstring: 'extra argument in signature',
          arguments: [
            new Argument({
              name: 'arg1',
              type: 'string',
              description: 'exists',
              required: true,
            }),
          ],
        });
      }).toThrow(/Argument names.*do not match/);
    });
  });

  describe('Function calling tool generation', () => {
    it('should generate OpenAI function calling tool', () => {
      const command = new Command({
        name: 'test_function',
        docstring: 'A test function for OpenAI',
        arguments: [
          new Argument({
            name: 'required_arg',
            type: 'string',
            description: 'Required string argument',
            required: true,
          }),
          new Argument({
            name: 'enum_arg',
            type: 'string',
            description: 'Enum argument',
            required: true,
            enum: ['option1', 'option2'],
          }),
          new Argument({
            name: 'optional_arg',
            type: 'integer',
            description: 'Optional integer argument',
            required: false,
          }),
        ],
      });

      const tool = command.getFunctionCallingTool();

      expect(tool.type).toBe('function');
      expect(tool.function.name).toBe('test_function');
      expect(tool.function.description).toBe('A test function for OpenAI');

      const properties = tool.function.parameters.properties;
      expect(properties.required_arg.type).toBe('string');
      expect(properties.optional_arg.type).toBe('integer');
      expect(properties.enum_arg.enum).toEqual(['option1', 'option2']);

      const required = tool.function.parameters.required;
      expect(required).toContain('required_arg');
      expect(required).toContain('enum_arg');
      expect(required).not.toContain('optional_arg');
    });
  });

  describe('Multi-line commands', () => {
    it('should handle commands with end markers', () => {
      const command = new Command({
        name: 'edit',
        signature: 'edit <filename>',
        docstring: 'Edit a file with multi-line content',
        endName: 'EOF',
        arguments: [
          new Argument({
            name: 'filename',
            type: 'string',
            description: 'file to edit',
            required: true,
          }),
        ],
      });

      expect(command.invokeFormat).toBe('edit {filename}');
      expect(command.endName).toBe('EOF');
    });
  });

  describe('Custom argument format', () => {
    it('should handle custom argument formatting', () => {
      const command = new Command({
        name: 'custom_format',
        docstring: 'Test custom argument formatting',
        arguments: [
          new Argument({
            name: 'arg1',
            type: 'string',
            description: 'Custom formatted argument',
            required: true,
            argumentFormat: '--{value}',
          }),
        ],
      });

      expect(command.arguments[0].argumentFormat).toBe('--{value}');
      expect(command.invokeFormat).toBe('custom_format {arg1} ');
    });
  });
});
